feat(ivy): introduce `select()` to incrementally invoke directive hostBindings and hooks

matsko opened the pull request.

mhevery commented on packages/core/src/render3/instructions/instructions.ts:
```diff
@@ -109,14 +109,22 @@ export function refreshDescendantViews(lView: LView) {
 
 /** Sets the host bindings for the current view. */
-export function setHostBindings(tView: TView, viewData: LView): void {
+let interimHostBindingsIndex = 0;
```
we store all of the global state in `state.ts` please move it there.

mhevery commented on packages/core/test/render3/integration_spec.ts:
```diff
@@ -618,7 +618,7 @@ describe('render3 integration test', () => {
        * {% } %}
        */
-      const TestCmpt = createComponent('test-cmpt', function(rf: RenderFlags, ctx: {value: any}) {
+      const TestCmpt = createComponent('test-cmpt', function (rf: RenderFlags, ctx: {value: any}) {
```
Why all this space changes. Do you have the right version of clang?

mhevery reviewed (commented):
few comments

benlesh commented on packages/core/src/render3/instructions/instructions.ts:
```diff
@@ -109,14 +109,20 @@ export function refreshDescendantViews(lView: LView) {
 /** Sets the host bindings for the current view. */
-export function setHostBindings(tView: TView, viewData: LView): void {
+export function setHostBindings(tView: TView, viewData: LView, slotIndex?: number): void {
+  const breakAtSlotIndex = slotIndex != null ? -slotIndex : null;
```
`slotIndex != null` will match a `slotIndex` of `0`. Not sure if that's intentional.

Also, can we add a comment about why we're inverting the number to be negative here? And maybe a dev mode assertion to make sure that any slotIndex that _is_ passed is within a certain expected range. (I could imagine someone doing an `indexOf` search or something (no idea why), getting a `-1`, then passing it in here and having crazy things happen.

benlesh commented on packages/core/src/render3/instructions/instructions.ts:
```diff
@@ -145,7 +151,11 @@ export function setHostBindings(tView: TView, viewData: LView): void {
     }
+
+    setInterimHostBindingsIndex(breakAtSlotIndex == null ? 0 : i);
+    setInterimExpandoStartIndex(breakAtSlotIndex == null ? 0 : bindingRootIndex);
```
More `==` comparisons here, should these be `===`? If we're worried about `undefined`, maybe we could just be more explicit or even coerce to `null` elsewhere?

benlesh commented on packages/core/src/render3/state.ts:
```diff
@@ -340,3 +340,21 @@ export function leaveView(newView: LView): void {
 }
+
+let interimHostBindingsIndex = 0;
+export function setInterimHostBindingsIndex(index: number) {
```
Needs documentation, mostly around intended use. (The name is pretty self-documenting as far as what it does)

benlesh commented on packages/core/src/render3/state.ts:
```diff
@@ -340,3 +340,21 @@ export function leaveView(newView: LView): void {
+  interimHostBindingsIndex = index;
+}
+
+export function getInterimHostBindingsIndex() {
```
Same as above. We should document intent.

benlesh commented on packages/core/src/render3/state.ts:
```diff
@@ -340,3 +340,21 @@ export function leaveView(newView: LView): void {
+}
+
+let interimExpandoStartIndex = 0;
+export function setInterimExpandoStartIndex(index: number) {
```
Same as above. We should document intent.

benlesh commented on packages/core/src/render3/state.ts:
```diff
@@ -340,3 +340,21 @@ export function leaveView(newView: LView): void {
+  interimExpandoStartIndex = index;
+}
+
+export function getInterimExpandoStartIndex() {
```
Same as above. We should document intent.

benlesh commented on packages/core/test/render3/host_binding_spec.ts:
```diff
@@ -1281,4 +1281,83 @@ describe('host bindings', () => {
         /* isAttribute */ false);
   });
+
+  it('should call host bindings element-by-element when select() is called', () => {
```
Should we be writing `TestBed` tests instead of `ComponentFixture`? @mhevery @kara?

kara replied:
Yep, we agreed as a team there would be no more handwritten generated code

benlesh reviewed (commented):
Just a few minor things, and then a question for @mhevery and @kara.

matsko committed:
feat(ivy): introduce `select()` to incrementally invoke directive hostBindings and hooks

Prior to this fix, all directive/component host bindings update code was
run AFTER the template function for the parent component was run. This
means that the associated host bindings for an element that has a
directive and/or component attached to it will not run inline with the
bindings on that element before the next element is processed. This
patch ensures that the all host bindings for an element are processed
before the next element is visited.

benlesh commented on packages/core/src/render3/instructions/instructions.ts:
```diff
@@ -105,20 +107,46 @@ export function refreshDescendantViews(lView: LView) {
-    for (let i = 0; i < tView.expandoInstructions.length; i++) {
+
+    let i = getInterimHostBindingsIndex();
+    while (i < tView.expandoInstructions.length) {
```
Seems like a `for` loop is still more idiomatic here, since `i` doesn't change anywhere else in the loop.

matsko replied:
It does change: https://github.com/angular/angular/pull/29486/files#diff-a66758223216005c664ce98175c4164fL126

benlesh replied:
Ah ha! In that case, I think `while` is better. Thanks for making that change.
